### PR TITLE
Always run validation inside the training loop epoch

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

The current training epoch structure follows the following pattern:
(`max_epochs=1, limit_train_batches=1, limit_val_batches=1`)

```python
'on_fit_start',
    # training
    'on_train_start',
        'on_epoch_start',
        'on_train_epoch_start',
            'on_train_batch_start',
            'on_train_batch_end',
        'on_train_epoch_end',
        'on_epoch_end',
        # intra-training validation
        'on_validation_start',
            'on_epoch_start',
            'on_validation_epoch_start',
                'on_validation_batch_start',
                'on_validation_batch_end',
            'on_validation_epoch_end',
            'on_epoch_end',
        'on_validation_end',
    'on_train_end',
'on_fit_end',
```

Pros:
- Clear distinction between `on_train_epoch_end` and `on_train_end`.

This means when validation is run at the end of the epoch (the classic loop structure), validation is considered to be out of the training scope. This is inconsistent with the pattern when validation is run in the middle of the epoch:

```python
'on_fit_start',
    # training
    'on_train_start',
        'on_epoch_start',
        'on_train_epoch_start',
            'on_train_batch_start',
            'on_train_batch_end',
            # intra-training validation
            'on_validation_start',
                'on_epoch_start',
                'on_validation_epoch_start',
                    'on_validation_batch_start',
                    'on_validation_batch_end',
                'on_validation_epoch_end',
                'on_epoch_end',
            'on_validation_end',
        'on_train_epoch_end',
        'on_epoch_end',
    'on_train_end',
'on_fit_end',
```

This PR changes the code so the later pattern is always used.

Pros:
- Consistent

Cons:
- Both training and validation dataloaders are in memory
- Messes up existing timing code
- Cannot checkpoint or early-stop on metrics logged `on_train_epoch_end` or inside `training_step` via `self.log(..., on_epoch=True)`. This is a breaking change. `EarlyStopping` users can set `check_on_train_epoch_end=True` to avoid the error. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -67,7 +67,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     * Refactored result handling in training loop ([#7506](https://github.com/PyTorchLightning/pytorch-lightning/pull/7506))
     * Moved attributes `hiddens` and `split_idx` to TrainLoop ([#7507](https://github.com/PyTorchLightning/pytorch-lightning/pull/7507))
     * Refactored the logic around manual and automatic optimization inside the optimizer loop ([#7526](https://github.com/PyTorchLightning/pytorch-lightning/pull/7526))
-
+    * Simplified "should run validation" logic ([#7682](https://github.com/PyTorchLightning/pytorch-lightning/pull/7682))
+    * Refactored "should run validation" logic when the trainer is signaled to stop ([#7701](https://github.com/PyTorchLightning/pytorch-lightning/pull/7701))
+    * Always run validation inside the training loop epoch ([#7357](https://github.com/PyTorchLightning/pytorch-lightning/pull/7357))
```
Comment (awaelchli) [inline CHANGELOG.md:72]: We cannot put this under refactor. A refactor would not change behavior, but this PR does and it is intentional. 
I would specifically mention the hook order changes in the changed section. In particular, if validation runs then the train_epoch_end hooks run after validation_end hooks

Comment (carmocca) [reply CHANGELOG.md:72]: Changing it.

The nuances of the change will be explained in the breaking changes section of the release notes.

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -529,44 +528,38 @@ def run_training_epoch(self):
...
             self.check_checkpoint_callback(True)
-
-        if should_check_val:
-            self.trainer.validating = True
-            self.trainer._run_evaluation(on_epoch=True)
-            self.trainer.training = True
-
-        if batch_output.signal != -1:
-            self.increment_accumulated_grad_global_step()
+            self.global_step += 1
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:562]: nit: could we add a TODO that this should be removed in favor of pushing this logic directly to the checkpoint callback?

**Review by ananthsub — APPROVED** [review]
thanks for the cleanup @carmocca !

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:562]: With "this" do you mean the `did_train_only` check, the `global_step` change, or both?

Comment (ananthsub) [reply pytorch_lightning/trainer/training_loop.py:562]: the whole `if did_train_only` force-calling the checkpoint callback

**Review by justusschock — APPROVED** [review]
I like it a lot!

Two questions to the pros/cons you mentioned:

* Having the data loader in memory shouldn't be a problem as before you also had both datasets in memory (the memory footprint of the loader itself is very low)
* Assuming you're talking about the worker processes: Shouldn't the train loader's worker processes vanish (of course only if set as non-persistent) as soon as they have finished iterating over the dataset (which would still be the case from my understanding)?

**Review by awaelchli — APPROVED** [review]
great stuff

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -529,44 +528,38 @@ def run_training_epoch(self):
...
+        self.global_step += 1
 
-        should_check_val = self._should_check_val_fx(batch_idx, is_last_batch, on_epoch=True)
-        should_skip_eval = self.trainer.evaluation_loop.should_skip_evaluation(self.trainer.num_val_batches)
-        should_train_only = self.trainer.disable_validation or should_skip_eval
-
-        # update epoch level lr_schedulers if no val loop outside train loop is triggered
-        if not should_check_val or should_train_only:
-            self.update_lr_schedulers('epoch')
+        self.update_lr_schedulers('epoch')
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:554]: Why? The logic hasn't changed.

Either we were running no validation and we had this call here under an `if` or we were running validation and we had the call in `_run_evaluation`

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:562]: Yes

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:562]: Answering @ananthsub 

> the whole if did_train_only force-calling the checkpoint callback

I'm not sure we want to move it inside the callback as it is logic tied to the loop structure.

**Comment by carmocca** [comment]
> Two questions to the cons you mentioned around the loader logic ...

I haven't checked if this is happening but what if your batches are huge and you keep the last train batch in GPU while you run validation? This is what I was referring to.

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:562]: > Yes

Then I think this is a great clean up.

I think what @ananthsub is saying is let the callback system call ModelCheckpoint. I think we should aim for that.

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:562]: Discussed offline with Adrian. Not going to add the TODO but will try to resolve this asap.

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:554]: it looked like the behavior changed
